less/_functions.less

```diff
@@ -3,7 +3,9 @@
 // Utility mixins and functions for evaluating source code across our variables, maps, and mixins.
 
 @plugin "plugins/index";
-@plugin "plugins/map-get";
+
+// Implements a keys function like map-keys
+@plugin "plugins/keys";
```

Comment:
Reverting back to the name `map-keys` for the time being just so we can get this PR merged.

On the subject of the aliasing problem though, I was thinking about it and it seems like a very simplified version of JavaScript’s module aliasing syntax could possibly work:

```less
@plugin "plugins/keys" as "map-keys";

// or maybe more like a media-query?

@plugin "plugins/keys" as (map-keys);
@plugin "plugins/keys" and (alias: map-keys);
```

It’s probably something that Less will need to address eventually as there could be naming conflicts between plugins.

Reply:
Wait, what is the need being expressed here?

Reply:
@seanCodes 

> as there could be naming conflicts between plugins.

That should never happen. If you're supplying a library, you should scope your plugin calls to a `{}` block so that they are not globally exposed.

Reply:
@matthew-dean If I create a Less plugin that just adds a function named `foo()` and then someone else creates a plugin that adds a function with the same name, one would override the other if both were imported, right?

Reply:
@seanCodes No, not really, and this is one of the big reasons that we de-emphasized "global" plugins in Less in favor of `@plugin`. IIRC, Less attaches an instance of a function registry to every ruleset. So, when you add a function, they're added to that local registry. In this way, functions are a bit like variables. The reason that most functions in Less are global is because they're added to the function registry that's attached to the root node. So, when you call a function, Less searches up the scope for a matching key. If none is found, it resolves to a plain text output of that function name.

Because of this, `@plugin` allows library authors to embed lots of extra functionality or JS-based processing that doesn't leak into the evaluation of other nodes. The previous generation of Less plugins didn't allow that, and if you needed JS functions for one specific `@import`-ed library, it would apply to the processing of all your Less code.

There are some exceptions to this. Plugins aren't restricted to just calling the Node tree and making global changes. So they're not "sandboxed" in any sense of the word, just with functions. Similarly, if you were to register a node visitor with `@plugin`, it would subsequently call that visitor for EVERY node (of that type), not just nodes within the scope. It's fundamentally a different type of problem to solve because visitors operate on an AST (so visits are outside in), whereas functions evaluate within a ruleset evaluation, so evaluation / lookup is inside out.

Does that make sense?

Reply:
@matthew-dean Got it, that makes a lot of sense, actually. Thanks for breaking that down!